### ISSUE TITLE
Text completion with packing?

### ISSUE THREAD

**Issue by kostum123** [huggingface/trl#1212]
Why we cant use packing with text completion training? Also can we use text completion training inside of the trl for teaching new information? Instead of finetuning model with instruction response pairs we finetune on raw corpus text.

**Comment by lvwerra** [comment]
Are you talking about masking the instructions? It would require a complicated attention mask to make sure we out the instructions of each example. In principle it is not quite clear yet if masking the instructions really helps and might depend a lot on the diversity of the instructions for example. And yes, you can use the `SFTTrainer` to fine-tune on raw text.

**Comment by kostum123** [comment]
> Are you talking about masking the instructions? It would require a complicated attention mask to make sure we out the instructions of each example. In principle it is not quite clear yet if masking the instructions really helps and might depend a lot on the diversity of the instructions for example. And yes, you can use the `SFTTrainer` to fine-tune on raw text.

I was considering both aspects. Indeed, I've learned that raw text fine-tuning is supported in TRL. I have a related question: suppose I aim to train the model on English quotes, like the example below. According to the TRL SFTTrainer documentation, dataset preprocessing, including packing, is automatically handled by SFTTrainer. Does this also mean that beginning-of-sentence (BOS), end-of-sentence (EOS), and padding tokens are automatically managed if we simply provide the necessary fields to the trainer? I attempted training with this logic, and the loss consistently decreased across multiple epochs; the model's loss was down to 0.6. Does this indicate successful learning? Also, if I disable packing, will the tokenization and handling of special tokens still occur?

Here's how I loaded the dataset and set up the trainer:

# Load the dataset
from datasets import load_dataset
dataset = load_dataset("Abirate/english_quotes", split="train")

# Set up the trainer
from trl import SFTTrainer
from transformers import TrainingArguments

trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    dataset_text_field="quote",
    max_seq_length=max_seq_length,
    packing=True,  # This option packs short sequences together to save time.
    args=TrainingArguments(
        per_device_train_batch_size=3,
        per_device_eval_batch_size=1,
        gradient_accumulation_steps=1,
        warmup_ratio=0.05,
        max_grad_norm=1.0,
        num_train_epochs=3,
        learning_rate=2e-5,
        fp16=not torch.cuda.is_bf16_supported(),
        bf16=torch.cuda.is_bf16_supported(),
        logging_steps=1,
        optim="paged_adamw_8bit",
        weight_decay=0.01,
        lr_scheduler_type="cosine",
        seed=3407,
        output_dir="outputs",
    ),
)
